### ISSUE TITLE
clang-offload-bundler mixes symbols from different objects when unbundling an archive

### ISSUE THREAD

**Issue by LU-JOHN** [llvm/llvm-project#83509]
Change to clang/lib/Driver/OffloadBundler.cpp in commit https://github.com/llvm/llvm-project/pull/81700 causes clang-offload-bundler to join symbols from different objects when unbundling an archive:

Source file s.cpp:
> void bazball() {
> } 

Source file t.cpp:
> void foobar() {
> } 

Commands:

> clang++ -c s.cpp  -o s.o;
> clang-offload-bundler -type=o  -targets=host-x86_64-unknown-linux-gnu -input=s.o  -output=s_bundled.o;
> clang++ -c t.cpp  -o t.o;
> clang-offload-bundler -type=o  -targets=host-x86_64-unknown-linux-gnu -input=t.o  -output=t_bundled.o;
> llvm-ar cr bundled.a s_bundled.o t_bundled.o
> clang-offload-bundler  -unbundle --targets=host-x86_64-unknown-linux-gnu -type=aoo -input=bundled.a -output=t_list.txt; nm \`cat t_list.txt\`

The output of the nm command without the change is:

> bundled-742762.o:
> 0000000000000000 T _Z7bazballv
> bundled-61506b.o:
> 0000000000000000 T _Z6foobarv 
> 

The output with the change is:

> /tmp/bundled-e7ca35.o:s_bundled.o:
> 0000000000000000 T _Z7bazballvt_bundled.o:
> 0000000000000000 T _Z6foobarv
> 
> /tmp/bundled-5ea078.o:s_bundled.o:
> 0000000000000000 T _Z7bazballvt_bundled.o:
> 0000000000000000 T _Z6foobarv

**Comment by jsji** [comment]
FYI. @yxsamliu

**Comment by LU-JOHN** [comment]
Sorry, I need to analyze this issue more carefully.  It depends on some non-community changes.

**Comment by LU-JOHN** [comment]
I have a concern about the code in commit https://github.com/llvm/llvm-project/pull/81700 at:

https://github.com/yxsamliu/llvm-project/blob/1e9c29c4794ed6d60cf54c902cd0f854afd1aace/clang/lib/Driver/OffloadBundler.cpp#L718

This code is used to get a processed bundle in an ObjectFileHandler by using objcopy to process BundlerConfig.InputFileNames.front().  My concern is how does this code guarantee that BundlerConfig.InputFileNames.front() is connected to this instance of ObjectFileHandler?   This caused a problem for me when unbundling an archive with multiple object files.   I hit this problem when using the command sequence in my earlier comment.  I have not been able to convert the command sequence into one that works with community tools, but I believe it should be possible when processing a multi-object archive.

**Comment by yxsamliu** [comment]
Good catch. The code assumes unbundling a non-archive file which has only one input. That is why it breaks for archive.

**Comment by yxsamliu** [comment]
Are you using some downstream llvm branch? Clang-offload-bundler in the trunk does not accept -type=aoo and also does not allow extract bundle entry for --targets=host-x86_64-unknown-linux-gnu

```
clang-offload-bundler -unbundle --targets=host-x86_64-unknown-linux-gnu -type=a -input=bundled.a -output=t_list.txt
clang-offload-bundler: error: no compatible code object found for the target 'host-x86_64-unknown-linux-gnu-' in heterogeneous archive library: bundled.a
```

**Comment by LU-JOHN** [comment]
Yes, it is an INTEL customization.   I was not able to figure out how to extract a multi-object archive with the official clang-offload-bundler.  Is it possible?
